### PR TITLE
Python-SDK Implementation for Ocean GKE

### PR DESCRIPTION
**Pull request by ramrutha497**
_This is the default pull request template. You can customize it by adding a `pull_request_template.md` at the root of your repo or inside the `.github` folder._

# Jira Ticket
https://spotinst.atlassian.net/browse/SPOTAUT-18355
https://spotinst.atlassian.net/browse/SPOTAUT-18443

_Include a link to your Jira Ticket_
_Example: [JIRAISS-1234](https://spotinst.atlassian.net/browse/JIRAISS-1234)_

# Demo

_Please add a recording of the feature/bug fix in work. if you added new routes, the recording should show the request and response for each new/changed route_

# Checklist:
- [ ] I have filled relevant self assessment ([NodeJS](https://docs.google.com/forms/d/e/1FAIpQLSfl14u9AOBAmxVJ272tvO7XNuXE-EMvWaGcaRZalu1UAKB7RA/viewform), [Frontend](https://docs.google.com/forms/d/e/1FAIpQLSdiBPNKH81w_EkavihVL8Uwb0j7tP8PwJmLFYm2nCOQxz-1qw/viewform), [Backend](https://docs.google.com/forms/d/e/1FAIpQLSed_PsTJ5-XIWkFL6BSDE2AQRBVPmwc3PAmHZkUn-erzVI37Q/viewform?usp=sf_link))
- [ ] I have run ESlint on my changes and fixed all warnings and errors (**NodeJS & Frontend Services**)
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have validated all the requirements in the Jira task were answered
- [ ] I have all neccessary approvals for the design/mini design of this task
- [ ] I have approved the API changes and granular permission patterns (documentation subtask) (**For public services only**)

### CODE REVIEW

File: spotinst_sdk2/clients/ocean/__init__.py [hunk 1]
```diff
@@ -1693,3 +1694,662 @@ def list_migrations(self, ocean_id: str):
...
+
+        formatted_response = self.convert_json(
+            response, self.camel_to_underscore)
+
+        return formatted_response["response"]["items"][0]
+
+    def update_ocean_cluster(self, ocean_id: str, ocean: gcp_ocean.Ocean):
+        """
+        All Ocean parameters are updatable, excluding the Name and controllerClusterId.
+        This API supports partial updates, so specific fields can be updated separately.
```
Comment (chandra1-n) [inline spotinst_sdk2/clients/ocean/__init__.py]: please change this to Update an existing Ocean Cluster

File: spotinst_sdk2/clients/ocean/__init__.py [hunk 2]
```diff
@@ -1693,3 +1694,662 @@ def list_migrations(self, ocean_id: str):
...
+            self.camel_to_underscore)
+
+        return formatted_response["response"]["items"][0]
+
+    def get_elastilog(self, ocean_id, from_date, to_date, severity=None, resource_id=None, limit=None):
+        """
+        Get group’s Elastilog by
+
+        # Arguments
+        to_date (String): to date
```
Comment (chandra1-n) [inline spotinst_sdk2/clients/ocean/__init__.py]: please modify to end date value

File: spotinst_sdk2/clients/ocean/__init__.py [hunk 3]
```diff
@@ -1693,3 +1694,662 @@ def list_migrations(self, ocean_id: str):
...
+        from_date (String): to date
```
Comment (chandra1-n) [inline spotinst_sdk2/clients/ocean/__init__.py]: please modify to beginning date value

File: spotinst_sdk2/clients/ocean/__init__.py [hunk 4]
```diff
@@ -1693,3 +1694,662 @@ def list_migrations(self, ocean_id: str):
...
+        severity(String) (Optional): Log level severity
+        resource_id(String) (Optional): Filter log extracted entires related to a
+          specific resource id
+        limit(String) (Optional): Maximum number of lines to extract in a response
```
Comment (chandra1-n) [inline spotinst_sdk2/clients/ocean/__init__.py]: limit is int not string

File: spotinst_sdk2/clients/ocean/__init__.py [hunk 5]
```diff
@@ -1693,3 +1694,662 @@ def list_migrations(self, ocean_id: str):
...
+
+    def get_elastilog(self, ocean_id, from_date, to_date, severity=None, resource_id=None, limit=None):
+        """
+        Get group’s Elastilog by
+
+        # Arguments
+        to_date (String): to date
+        from_date (String): to date
+        severity(String) (Optional): Log level severity
+        resource_id(String) (Optional): Filter log extracted entires related to a
```
Comment (chandra1-n) [inline spotinst_sdk2/clients/ocean/__init__.py]: please correct the description

File: spotinst_sdk2/clients/ocean/__init__.py [hunk 6]
```diff
@@ -1693,3 +1694,662 @@ def list_migrations(self, ocean_id: str):
...
+
+        excluded_missing_dict = self.exclude_missing(
+            json.loads(aggregated_cluster_costs_request.toJSON()))
+
+        formatted_missing_dict = self.convert_json_with_list_of_lists(
+            excluded_missing_dict, self.underscore_to_camel)
+
+        body_json = json.dumps(formatted_missing_dict)
+
+        aggregated_costs_response = self.send_post(
```
Comment (chandra1-n) [inline spotinst_sdk2/clients/ocean/__init__.py]: modify the variable name to aggregated_summary_costs_response

File: spotinst_sdk2/clients/ocean/__init__.py [hunk 7]
```diff
@@ -1693,3 +1694,662 @@ def list_migrations(self, ocean_id: str):
...
+            body=body_json,
+            url=self.__base_ocean_cluster_url + "/" + ocean_id + "/aggregatedCosts/summary",
+            entity_name='ocean (aggregated cluster costs)')
```
Comment (chandra1-n) [inline spotinst_sdk2/clients/ocean/__init__.py]: modify the entity name to display summary costs

File: spotinst_sdk2/clients/ocean/__init__.py [hunk 8]
```diff
@@ -1693,3 +1694,662 @@ def list_migrations(self, ocean_id: str):
...
+
+        return formatted_response["response"]["items"][0]
+
+    def create_virtual_node_group(self, vng: gcp_ocean.VirtualNodeGroup):
+        """
+        Create a virtual node group.
+
+        # Arguments
+        vng (VirtualNodeGroup): VirtualNodeGroup Object
+        initial_nodes: When set to an integer greater than 0, a corresponding number of nodes will be launched from the virtual node group created.
```
Comment (chandra1-n) [inline spotinst_sdk2/clients/ocean/__init__.py]: this attribute is not part of the query param

File: spotinst_sdk2/clients/ocean/__init__.py [hunk 9]
```diff
@@ -1693,3 +1694,662 @@ def list_migrations(self, ocean_id: str):
...
+            body=body_json,
+            url=self.__base_ocean_launchspec_url,
+            entity_name='ocean_gcp_vng')
+
+        formatted_response = self.convert_json(response,
+                                               self.camel_to_underscore)
+
+        return formatted_response["response"]["items"][0]
+
+    def get_all_ocean_vngs(self, ocean_id: str):
```
Comment (chandra1-n) [inline spotinst_sdk2/clients/ocean/__init__.py]: please modify the name to get_all_virtual_node_groups

File: spotinst_sdk2/clients/ocean/__init__.py [hunk 10]
```diff
@@ -1693,3 +1694,662 @@ def list_migrations(self, ocean_id: str):
...
+        )
+
+        formatted_response = self.convert_json(
+            response, self.camel_to_underscore)
+
+        return formatted_response["response"]["items"]
+
+    def import_vng_configuration(self, node_pool_name: str, ocean_id: str):
+        """
+        Import cluster configuration of an AKS cluster to use in create_ocean_cluster api call
```
Comment (chandra1-n) [inline spotinst_sdk2/clients/ocean/__init__.py]: please modify the description accordingly as it it mentioning about the AKS cluster but the function is to import vng

File: spotinst_sdk2/clients/ocean/__init__.py [hunk 11]
```diff
@@ -1693,3 +1694,662 @@ def list_migrations(self, ocean_id: str):
...
+
+        # Returns
+        (Object): Ocean API response
+        """
+
+        response = self.send_post_with_params(
+            body=None,
+            url=self.__base_ocean_launchspec_url + "/import",
+            entity_name='ocean_gcp',
```
Comment (chandra1-n) [inline spotinst_sdk2/clients/ocean/__init__.py]: entity_name should be ocean_gcp_vng?

File: spotinst_sdk2/clients/ocean/__init__.py [hunk 12]
```diff
@@ -1693,3 +1694,662 @@ def list_migrations(self, ocean_id: str):
...
+
+        formatted_missing_dict = self.convert_json(
+            excluded_missing_dict, self.underscore_to_camel)
+
+        body_json = json.dumps(formatted_missing_dict)
+
+        response = self.send_put(
+            body=body_json,
+            url=self.__base_ocean_launchspec_url + "/" + vng_id,
+            entity_name='ocean_aws_vng')
```
Comment (chandra1-n) [inline spotinst_sdk2/clients/ocean/__init__.py]: wrong entity_name

File: spotinst_sdk2/clients/ocean/__init__.py [hunk 13]
```diff
@@ -1693,3 +1694,662 @@ def list_migrations(self, ocean_id: str):
...
+
+        # Arguments
+        ocean_launch_spec_id (String): Ocean cluster launch specification identifier
+
+        # Returns
+        (Object): Ocean Allowed Instance Types response
+        """
+        response = self.send_get(
+            url=self.__base_ocean_launchspec_url + "/" + ocean_launch_spec_id,
+            entity_name="ocean"
```
Comment (chandra1-n) [inline spotinst_sdk2/clients/ocean/__init__.py]: entity_name should be ocean_gcp_vng

File: spotinst_sdk2/models/ocean/gcp/__init__.py [hunk 14]
```diff
@@ -0,0 +1,984 @@
...
+            target: int = none
+            ):
+        self.maximum = maximum
+        self.minimum = minimum
+        self.target = target
+# endregion
+
+
+# region Compute
+class LocationType(Enum):
```
Comment (chandra1-n) [inline spotinst_sdk2/models/ocean/gcp/__init__.py:119]: did not find this in document

File: spotinst_sdk2/models/ocean/gcp/__init__.py [hunk 15]
```diff
@@ -0,0 +1,984 @@
...
+        self.enable_integrity_monitoring = enable_integrity_monitoring
+        self.enable_secure_boot = enable_secure_boot
+
+
+class LaunchSpecification:
+    """
+    # Arguments
+    ip_forwarding: bool
+    labels: List[Labels]
+    metadata: List[Metadata]
```
Comment (chandra1-n) [inline spotinst_sdk2/models/ocean/gcp/__init__.py:230]: please re-verify in api, it seems it's an object not array of abject

File: spotinst_sdk2/models/ocean/gcp/__init__.py [hunk 16]
```diff
@@ -0,0 +1,984 @@
...
+        self.memory_per_unit = memory_per_unit
+        self.num_of_units = num_of_units
+
+
+class AutoScale:
+    """
+    # Arguments
+    auto_headroom_percentage: int
+    headrooms: List[Headroom]
+    down: Down
```
Comment (chandra1-n) [inline spotinst_sdk2/models/ocean/gcp/__init__.py:715]: no definition for Down object found

File: spotinst_sdk2/models/ocean/gcp/__init__.py [hunk 17]
```diff
@@ -0,0 +1,984 @@
...
+
+
+class Compute:
+    """
+    # Arguments
+    availability_zones: List[str]
+    backend_services: BackendServices
+    instance_types: InstanceTypes
+    launch_specification: LaunchSpecification
+    network_interfaces: List(NetworkInterfaces)
```
Comment (chandra1-n) [inline spotinst_sdk2/models/ocean/gcp/__init__.py]: it should be List[NetworkInterfaces]

**Review by chandra1-n — CHANGES_REQUESTED** [review]
please fix the review comments

File: spotinst_sdk2/clients/ocean/__init__.py (hunk 1 above)
Comment (ramrutha497) [reply spotinst_sdk2/clients/ocean/__init__.py]: Done

File: spotinst_sdk2/clients/ocean/__init__.py (hunk 2 above)
Comment (ramrutha497) [reply spotinst_sdk2/clients/ocean/__init__.py]: Done

File: spotinst_sdk2/clients/ocean/__init__.py (hunk 3 above)
Comment (ramrutha497) [reply spotinst_sdk2/clients/ocean/__init__.py]: Done

File: spotinst_sdk2/clients/ocean/__init__.py (hunk 4 above)
Comment (ramrutha497) [reply spotinst_sdk2/clients/ocean/__init__.py]: Done

File: spotinst_sdk2/clients/ocean/__init__.py (hunk 5 above)
Comment (ramrutha497) [reply spotinst_sdk2/clients/ocean/__init__.py]: Done

File: spotinst_sdk2/clients/ocean/__init__.py (hunk 6 above)
Comment (ramrutha497) [reply spotinst_sdk2/clients/ocean/__init__.py]: Done

File: spotinst_sdk2/clients/ocean/__init__.py (hunk 7 above)
Comment (ramrutha497) [reply spotinst_sdk2/clients/ocean/__init__.py]: Done

File: spotinst_sdk2/clients/ocean/__init__.py (hunk 8 above)
Comment (ramrutha497) [reply spotinst_sdk2/clients/ocean/__init__.py]: Done

File: spotinst_sdk2/clients/ocean/__init__.py (hunk 9 above)
Comment (ramrutha497) [reply spotinst_sdk2/clients/ocean/__init__.py]: Done

File: spotinst_sdk2/clients/ocean/__init__.py (hunk 10 above)
Comment (ramrutha497) [reply spotinst_sdk2/clients/ocean/__init__.py]: Done

File: spotinst_sdk2/clients/ocean/__init__.py (hunk 11 above)
Comment (ramrutha497) [reply spotinst_sdk2/clients/ocean/__init__.py]: Done

File: spotinst_sdk2/clients/ocean/__init__.py (hunk 12 above)
Comment (ramrutha497) [reply spotinst_sdk2/clients/ocean/__init__.py]: Done

File: spotinst_sdk2/clients/ocean/__init__.py (hunk 13 above)
Comment (ramrutha497) [reply spotinst_sdk2/clients/ocean/__init__.py]: Done

File: spotinst_sdk2/models/ocean/gcp/__init__.py (hunk 17 above)
Comment (ramrutha497) [reply spotinst_sdk2/models/ocean/gcp/__init__.py]: Done

File: spotinst_sdk2/clients/ocean/__init__.py [hunk 18]
```diff
@@ -1693,3 +1694,661 @@ def list_migrations(self, ocean_id: str):
...
+            body=body_json,
+            url=self.__base_ocean_cluster_url,
+            entity_name='ocean')
+
+        formatted_response = self.convert_json(response,
+                                               self.camel_to_underscore)
+
+        return formatted_response["response"]["items"][0]
+
+    def get_all_ocean_cluster(self):
```
Comment (chandra1-n) [inline spotinst_sdk2/clients/ocean/__init__.py]: modify to get_all_ocean_clusters

File: spotinst_sdk2/clients/ocean/__init__.py [hunk 19]
```diff
@@ -1693,3 +1694,661 @@ def list_migrations(self, ocean_id: str):
...
+            url=self.__base_ocean_cluster_url + "/" + ocean_id + "/roll",
+            entity_name="ocean (Cluster Roll)"
+        )
+
+        formatted_response = self.convert_json(
+            response, self.camel_to_underscore)
+
+        return formatted_response["response"]["items"]
+
+    def update_roll(self, ocean_id: str, roll_id: str, status: str):
```
Comment (chandra1-n) [inline spotinst_sdk2/clients/ocean/__init__.py:2176]: status is required according to optimizer code, please change

File: spotinst_sdk2/models/ocean/gcp/__init__.py [hunk 20]
```diff
@@ -0,0 +1,984 @@
...
+    availability_zones: List[str]
+    instance_types: List[str]
+    labels: List[Labels]
+    metadata: List[Metadata]
+    name: str
+    network_interfaces: List[NetworkInterfaces]
+    ocean_id: str
+    resource_limits: ResourceLimits
+    restrict_scale_down: bool
+    root_volume_size: int
```
Comment (chandra1-n) [inline spotinst_sdk2/models/ocean/gcp/__init__.py]: modify to root_volume_size_in_gb

**Review by chandra1-n — CHANGES_REQUESTED** [review]
please fix the review comments

File: spotinst_sdk2/models/ocean/gcp/__init__.py (hunk 16 above)
Comment (ramrutha497) [reply spotinst_sdk2/models/ocean/gcp/__init__.py:715]: Used from cluster object

File: spotinst_sdk2/clients/ocean/__init__.py (hunk 18 above)
Comment (ramrutha497) [reply spotinst_sdk2/clients/ocean/__init__.py]: Done

File: spotinst_sdk2/models/ocean/gcp/__init__.py (hunk 20 above)
Comment (ramrutha497) [reply spotinst_sdk2/models/ocean/gcp/__init__.py]: Done

**Review by chandra1-n — DISMISSED** [review]

**Review by chandra1-n — APPROVED** [review]

**Pull request merged**